src/azul/openapi/validation.py

```diff
@@ -0,0 +1,144 @@
+import json
+import logging
+from typing import (
+    Mapping,
+)
+
+from chalice import (
+    BadRequestError,
+)
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul.openapi.schema import (
+    type_lookup,
+)
+from azul.uuids import (
+    InvalidUUIDError,
+    InvalidUUIDVersionError,
+    validate_uuid,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SpecValidator:
+
+    def __init__(self,
+                 content: Mapping = None,
+                 pattern: str = None,
+                 name: str = None,
+                 required: bool = None,
+                 schema: Mapping = None,
+                 **kwargs):
+        """
+        :param kwargs: Since in is a Python reserved word we can only extract it
+         using kwargs. All other kwargs are ignored.
+        """
+        self.content = content
+        self.pattern = pattern
+        self.name = name
+        self.required = required
+        self.schema = schema
+        if self.schema is not None and self.content is not None:
+            raise TypeError('OpenAPI spec enforces that schema and content are mutually exclusive')
+        if self.schema:
+            parameter_schema = self.schema
+        else:
+            for k, v in self.content.items():
+                parameter_schema = v['schema']
+        self.validator = type_lookup(parameter_schema['type'])
+        self.json_validators = dict(jsonschema.Draft4Validator.VALIDATORS)
+        self._in = kwargs['in']
+
+    def validate(self, request_parameter: Mapping[str, str]):
+        """
+        Validates if a given request parameter meets the spec that is
+        specified for the endpoint.
+        :param request_parameter: A dict with the keys 'name' and 'value'.
+        """
+
+        param_name = request_parameter['name']
+        param_value = request_parameter['value']
+
+        if param_name == 'filters':
+            self.schema = self.content['application/json']['schema']
+            try:
+                filter_json = json.loads(param_value)
+            except json.decoder.JSONDecodeError:
+                raise BadRequestError(f'The {param_name!r} parameter is not valid JSON')
+            if type(filter_json) is not dict:
+                raise BadRequestError(f'The {param_name!r} parameter must be a JSON object')
+            else:
+                for facet, filter_ in filter_json.items():
+                    try:
+                        relation, value = one(filter_.items())
+                    except ValueError:
+                        raise BadRequestError(f"The 'filters' parameter entry for {facet!r} may only "
+                                              f"specify a single relation")
```

Comment:
I removed the f in `f"..."` because it's not necessary on this line. Same thing with line 83.

Reply:
although its not necessary, im not a fan of how it looks when the 'f' is removed on some lines and not others.
I brought this up with Hannes his response was,
 @hannes-ucsc "f on every line"